### PR TITLE
Implementation of ct_select intrinsic frontend

### PR DESCRIPTION
**Pull request by kumarak**
Clang frontend changes to support ct.select intrinsic and update codegen pipeline to emit code for intrinsic.

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp [hunk 1]
```diff
@@ -6659,6 +6659,72 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
...
+
+    if (TLI.isSelectSupported(
+            TargetLoweringBase::SelectSupportKind::CtSelect)) {
+      SDValue Result = DAG.getNode(ISD::CTSELECT, DL, VT, Cond, A, B);
+      setValue(&I, Result);
+      return;
+    }
+
+    llvm::errs() << "intrinsic ct_select not supported\n";
+    SDValue Result = DAG.getNode(ISD::SELECT, DL, VT, Cond, A, B);
```
Comment (kaoudis) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:6687]: To check my understanding, this means for targets that don't support the built-in, we fall back to "regular" select?

File: clang/test/Sema/builtin-ct-select-edge-cases.c [hunk 2]
```diff
@@ -0,0 +1,384 @@
+// RUN: %clang_cc1 -fsyntax-only -verify %s
+// RUN: %clang_cc1 -fsyntax-only -verify %s -fexperimental-new-constant-interpreter
+
```
Comment (kaoudis) [inline clang/test/Sema/builtin-ct-select-edge-cases.c:3]: Couple more edge case suggestions we may want to support / test for (FFTI if not! also apologies if we HAVE any of these and I'm missing them):

- nested __builtin_ct_select() (with a ctselect as the `cond`, or `a`, or `b`)
- `a` or `b` (hopefully both?) are vectors (I think we talked about `cond` being a vector not being supported yet?)
- what happens when cond evaluation is in UB territory? not sure how I would write this, maybe another NaN case, or the result of division by zero is cond or something?
- anything i386 specific that is a bit quirky that we haven't covered yet? I'm not sure what that would be other than lack of avx, or lack of 64-bit.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 3]
```diff
@@ -12062,6 +12066,118 @@ SDValue DAGCombiner::visitSELECT(SDNode *N) {
...
+    // X, Y
+    if (N2->getOpcode() == ISD::CTSELECT && N2->hasOneUse()) {
+      SDValue N2_0 = N2->getOperand(0);
+      SDValue N2_1 = N2->getOperand(1);
+      SDValue N2_2 = N2->getOperand(2);
+      if (N2_1 == N1 && N0.getValueType() == N2_0.getValueType()) {
+        // Create the actual or node if we can generate good code for it.
+        if (!normalizeToSequence) {
+          SDValue Or = DAG.getNode(ISD::OR, DL, N0.getValueType(), N0, N2_0);
+          return DAG.getNode(ISD::CTSELECT, DL, N1.getValueType(), Or, N1, N2_2,
```
Comment (kaoudis) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:12164]: Could this (and any other returns of ISD::CTSELECT in the lowering fn) infinitely recurse?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 4]
```diff
@@ -12062,6 +12066,118 @@ SDValue DAGCombiner::visitSELECT(SDNode *N) {
...
+                             Flags);
+        }
+        // Otherwise see if we can optimize to a better pattern.
+        if (SDValue Combined = visitORLike(N0, N2_0, DL))
+          return DAG.getNode(ISD::CTSELECT, DL, N1.getValueType(), Combined, N1,
```
Comment (kaoudis) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:12169]: Same as above - maybe you've got a clear picture of how this and the other return of the CTSELECT will/do work and I'm just not seeing it yet?

File: llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp [hunk 5]
```diff
@@ -5248,7 +5257,8 @@ void SelectionDAGLegalize::PromoteNode(SDNode *Node) {
     Results.push_back(DAG.getNode(ISD::TRUNCATE, dl, OVT, Tmp2));
     break;
   }
-  case ISD::SELECT: {
+  case ISD::SELECT:
+  case ISD::CTSELECT: {
```
Comment (kaoudis) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:5261]: To confirm I understand correctly, this yields effectively a final fallback to ISD::SELECT handling?

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 6]
```diff
@@ -24901,6 +24930,316 @@ static SDValue LowerSIGN_EXTEND_Mask(SDValue Op, const SDLoc &dl,
   return V;
 }
 
+/// Helper function to create a mask for LSB operations
+static SDValue createLSBMask(SDValue Input, EVT SplatVT, EVT CmpVT, SDLoc DL,
```
Comment (kaoudis) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:24934]: should this helper have ctselect in the naming? or are we cool with people using it for other purposes? I'm fine either way just curious

**Review by kaoudis — APPROVED** [review]
Just a handful of random questions really! Thanks for talking about this in Slack and aside from edge case-y stuff and what we already talked about (avx, vector cond type), looks good to me, pending Francesco's agreement.

File: clang/test/Sema/builtin-ct-select-edge-cases.c (hunk 2 above)
Comment (kaoudis) [reply clang/test/Sema/builtin-ct-select-edge-cases.c:3]: I think I just saw nested builtins, but for `a` being the nested internal builtin. What if `cond` was also a ctselect?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 7]
```diff
@@ -12062,6 +12066,118 @@ SDValue DAGCombiner::visitSELECT(SDNode *N) {
...
+  SDValue N0 = N->getOperand(0);
+  SDValue N1 = N->getOperand(1);
+  SDValue N2 = N->getOperand(2);
+  EVT VT = N->getValueType(0);
+  EVT VT0 = N0.getValueType();
+  SDLoc DL(N);
+  SDNodeFlags Flags = N->getFlags();
+
+  // if (SDValue V = DAG.simplifySelect(N0, N1, N2))
+  //   return V;
```
Comment (frabert) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:12079]: I'd personally remove commented-out code

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 8]
```diff
@@ -12062,6 +12066,118 @@ SDValue DAGCombiner::visitSELECT(SDNode *N) {
...
+
+  // ctselect (not Cond), N1, N2 -> ctselect Cond, N2, N1
+  if (SDValue F = extractBooleanFlip(N0, DAG, TLI, false)) {
+    SDValue SelectOp = DAG.getNode(ISD::CTSELECT, DL, VT, F, N2, N1);
+    SelectOp->setFlags(Flags);
+    return SelectOp;
+  }
+
+  // if (SDValue V = foldSelectOfConstants(N))
+  //   return V;
```
Comment (frabert) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:12092]: ditto re: commented-out code

**Review by frabert — CHANGES_REQUESTED** [review]
LGTM other than for a couple of stylistic changes :)

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp (hunk 1 above)
Comment (kumarak) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:6687]: Yes, if the target does not support `CTSELECT`, It will fallback to ISD::SELECT and follow the codegen.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp (hunk 4 above)
Comment (kumarak) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:12169]: May be I should remove optimizing ctselect to `and (zext cond), op0`. We may loose control over conditional select and what if future lowering optimizes it to branches.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp (hunk 3 above)
Comment (kumarak) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:12164]: These are select chaining optimizations that are done based on the OperandType. The combine of select condition only takes place when it is `ExpandInteger`, `ExpandFloat` or `SplitVector` types. This type-based normalization can avoid the case of infinite recursion. 
These are also optimization and we can disable them if we think there could be side-effect .

File: llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp (hunk 5 above)
Comment (kumarak) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:5261]: Yes, I believe type promotion is the last step of Selection DAG before hooking to target lowering.

File: clang/test/Sema/builtin-ct-select-edge-cases.c (hunk 2 above)
Comment (kumarak) [reply clang/test/Sema/builtin-ct-select-edge-cases.c:3]: I added unit tests with undefined behaviour. This is causing some of the unit tests to fail. Will look into them.

**Review by frabert — APPROVED** [review]

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 6 above)
Comment (kumarak) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:24934]: Inlined this function to avoid this helper function. Lowering function was too big so added to refactor it.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp (hunk 3 above)
Comment (kaoudis) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:12164]: Gotcha! I don't know if there could be side effects, mainly asking to rectify my ignorance about the bigger picture of the DAG combiner :D thank you.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp (hunk 4 above)
Comment (kaoudis) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:12169]: Which optimisations are you worried about in this case w/r/t the `and (zext cond), op0`?

Comment (kumarak) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:12169]: I have not looked into any specific optimization causing w/r/t `and (zext cond), op0` but want to make sure `ctselect` does not get optimized to any other instruction before target lowering.
